x/ibc-rate-limit/ibc_middleware_test.go

```diff
@@ -0,0 +1,275 @@
+	suite.Require().Equal(used, half.MulRaw(2))
+
+	// Sending above the quota should fail.
+	suite.AssertSendSuccess(false, suite.NewValidMessage(true, sdk.NewInt(1)))
+	return attrs
+}
+
+// Test rate limits are reset when the specified time period has passed
+func (suite *MiddlewareTestSuite) TestSendTransferReset() {
+	// Same test as above, but the quotas get reset after time passes
+	attrs := suite.TestSendTransferWithRateLimiting()
+	parts := strings.Split(attrs["weekly_period_end"], ".") // Splitting timestamp into secs and nanos
+	secs, _ := strconv.ParseInt(parts[0], 10, 64)
+	nanos, _ := strconv.ParseInt(parts[1], 10, 64)
+	resetTime := time.Unix(secs, nanos)
+
+	// Move both chains one block
+	suite.chainA.NextBlock()
+	suite.chainA.SenderAccount.SetSequence(suite.chainA.SenderAccount.GetSequence() + 1)
+	suite.chainB.NextBlock()
+	suite.chainB.SenderAccount.SetSequence(suite.chainB.SenderAccount.GetSequence() + 1)
+
+	// Reset time + one second
+	oneSecAfterReset := resetTime.Add(time.Second)
+	suite.coordinator.IncrementTimeBy(oneSecAfterReset.Sub(suite.coordinator.CurrentTime))
+
+	// Sending should succeed again
+	suite.AssertSendSuccess(true, suite.NewValidMessage(true, sdk.NewInt(1)))
+}
+
+// Test rate limiting on receives
+func (suite *MiddlewareTestSuite) TestRecvTransferWithRateLimiting() {
+	// Setup contract
+	suite.chainA.StoreContractCode(&suite.Suite)
+	quotas := suite.BuildChannelQuota("weekly", 604800, 5, 5)
+	addr := suite.chainA.InstantiateContract(&suite.Suite, quotas)
+	suite.chainA.RegisterRateLimitingContract(addr)
+
+	// Setup receiver chain's quota
+	osmosisApp := suite.chainA.GetOsmosisApp()
+
+	// Each user has 10% of the supply
+	supply := osmosisApp.BankKeeper.GetSupply(suite.chainA.GetContext(), sdk.DefaultBondDenom)
+	quota := supply.Amount.QuoRaw(20)
+	half := quota.QuoRaw(2)
+
+	// receive 2.5% (quota is 5%)
+	suite.AssertReceiveSuccess(true, suite.NewValidMessage(false, half))
+
+	// receive 2.5% (quota is 5%)
+	suite.AssertReceiveSuccess(true, suite.NewValidMessage(false, half))
+
+	// Sending above the quota should fail. Adding some extra here because the cap is increasing. See test bellow.
+	suite.AssertReceiveSuccess(false, suite.NewValidMessage(false, sdk.NewInt(1)))
+}
+
+// Test no rate limiting occurs when the contract is set, but not quotas are condifured for the path
+func (suite *MiddlewareTestSuite) TestSendTransferNoQuota() {
+	// Setup contract
+	suite.chainA.StoreContractCode(&suite.Suite)
+	addr := suite.chainA.InstantiateContract(&suite.Suite, ``)
+	suite.chainA.RegisterRateLimitingContract(addr)
+
+	// send 1 token.
+	// If the contract doesn't have a quota for the current channel, all transfers are allowed
+	suite.AssertSendSuccess(true, suite.NewValidMessage(true, sdk.NewInt(1)))
+}
+
+// Test the contract used for these tests is the same contract used for E2E testing
+func (s *MiddlewareTestSuite) TestRateLimitingE2ETestsSetupCorrectly() {
```

Comment:
Wdyt about importing the contract from a single location from both e2e and here?

Reply:
I was discussing that with @czarcas7ic. I'll probably change that so that the e2e tests copy the contract from the ibc-rate-limit module before they run. 

I don't like the idea of e2e needing to modify the fs, but I don't see many alternatives. Right now I'm also adding code that writes a param change json spec to the fs when running e2e tests, which I wish there was a better solution for

Reply:
I might not be having the full context here but why is e2e modifying the fs a concern for using the same contract path?